core/classifiers/CodeClassifier/CodeClassifier.py

```diff
@@ -0,0 +1,663 @@
+        list(str). A list containing tokens of input program.
+    """
+    token_program = []
+    for token_id, token_name in get_token(program):
+        # Ignore all newline, comment and empty string tokens.
+        if (token_id == token.N_TOKENS or token_id == 54
+                or token_name.strip() == ''):
+            continue
+        elif token_id == token.NAME:
+            # If token_id is tokens.NAME then only add if it is a python
+            # keyword. Treat all variables and methods same.
+            if token_name in keyword.kwlist:
+                token_program.append(token_name)
+            else:
+                token_program.append(var_token)
+        else:
+            token_program.append(token_name)
+
+    return token_program
+
+
+def generate_token_to_id(
+        data, threshold, var_token=VAR_TOKEN, unk_token=UNK_TOKEN):
+    """Generates a list of valid tokens and assigns a unique ID to each
+    token.
+
+    Args:
+        data: dict. A dictionary containing training_data.
+        threshold: int. Ignore a token if it appears less than 'threshold' times
+            in dataset.
+        var_token: str. Replace variable and method names with this string when
+            tokenizing data.
+        unk_token: ste. Replace unknown tokens (ignored tokens) with this string
+            when tokenizing the data.
+
+    Returns:
+        dict. A dictionary containing tokens mapped to a unique ID.
+    """
+    # All unique tokens and number of time they occur in dataset. A token
+    # can be a keyword, an identifier, a constant, an operator.
+    alphabet = {}
+
+    for pid in data:
+        program = data[pid]['source']
+        for token_id, token_name in get_token(program):
+            # Ignore all newline, comment and empty string tokens.
+            if (token_id == token.N_TOKENS or token_id == 54
+                    or token_name.strip() == ''):
+                continue
+            # If token_id is tokens.NAME then only add if it is a python
+            # keyword.
+            elif token_id == token.NAME:
+                if token_name in keyword.kwlist:
+                    alphabet[token_name] = alphabet.get(token_name, 0) + 1
+            else:
+                alphabet[token_name] = alphabet.get(token_name, 0) + 1
+
+    # Consider only those tokens which occur for more than threshold times
+    # in entire dataset.
+    valid_tokens = [k for k, v in alphabet.iteritems() if v > threshold]
+    token_to_id = dict(zip(valid_tokens, xrange(0, len(valid_tokens))))
+
+    # Add 'UNK' in token_to_id. This will be used to replace any token
+    # occurring in program which is not in valid_token.
+    token_to_id[unk_token] = len(token_to_id)
+
+    # Add 'V' in token_to_id. This token will be used to replace all
+    # variables and methods in program.
+    token_to_id[var_token] = len(token_to_id)
+    return token_to_id
+
+
+def tokenize_data(data, threshold=5, var_token=VAR_TOKEN, unk_token=UNK_TOKEN):
```

Comment:
Perhaps, we can make all the pre-processing steps re-usable? I agree that GridSearch and sample weight will not move out, in fact, I'm not even talking about anything inside the train() method. That is fine as it is. So, I'm ok with keeping SVM specific things in one file but at least think about creating a separate reusable module for winnowing. At least, that shouldn't be hard because you have a specific input for winnowing and you expect a specific output from it (there is no other dependency with SVM or anything).